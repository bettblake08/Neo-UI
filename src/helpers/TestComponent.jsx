export default {
  setState: jest.fn(),
  state: {}
}
