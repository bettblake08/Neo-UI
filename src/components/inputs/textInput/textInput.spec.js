import React from 'react';
import { shallow, mount } from "enzyme";
import waitForTest from "wait-for-expect";
import TextInput from "./textInput";
import TestComponent from "../../../helpers/TestComponent";
import Regex from "../../../helpers/regex";

describe("TextInput ", () => {
  const defaultProps = {
    name: "Name",
    defaultStatus: 0,
    parent: TestComponent,
    config: {
      text: "",
      floatingLabel: true,
      label: "Name",
      type: "neo-text-input",
      placeholder: "Brian Bett",
      length: 30
    }
  };

  /**
   * Sets up a enzyme wrapper
   * @param object PropsOverride A object of values used to override the props passed into the component being tested
   * @param boolean shouldMount A bool value to select which type of wrapper to create
   * @return object { wrapper, props }
  */
  const setup = (propsOverride, shouldMount = false) => {
    const newProps = { ...{ ...defaultProps, ...propsOverride } };
    const wrapper = shouldMount ? mount(<TextInput {...newProps} />) : shallow(<TextInput {...newProps} />);
    return { wrapper, props: newProps };
  }

  /** 
   * Tests the text input according to different scenarios
   * @param object testCase Contains values for a specified scenario
   * @param function testInputSpy A mock function to test an input check by using a function
  */
  const testTextInput = async (testCase, testInputSpy = null) => {
    const { wrapper } = setup({ config: { ...defaultProps.config, testInput: testCase.testInput } });
    const input = wrapper.find('input');
    input.simulate('change', { target: { value: testCase.value} });
    await waitForTest(() => {
      if (testInputSpy) expect(testInputSpy).toHaveBeenCalled();
      expect(wrapper.state('status')).toBe(testCase.status);
      expect(wrapper.instance().isValid()).toBe(testCase.status === 5);
    });
  }

  /** 
   * Tests the set status method according to different scenarios
   * @param object testCase Contains the values for a specified scenario
  */
  const testSetStatusMethod = testCase => {
    const { wrapper } = setup({ config: { ...defaultProps.config } });
    wrapper.instance().setStatus(...testCase.params);
    expect(wrapper.state('status')).toBe(testCase.statusNo);
  }

  it("renders as expected", () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders with minimal config props", async () => {
    const { wrapper } = setup({
      config: {
        ...defaultProps.config,
        length: null,
        label: null,
        placeholder: null,
        floatingLabel: false
      }
    }, true);
    expect(wrapper).toMatchSnapshot();
  });

  it("calls the onChange action when function test is valid", async () => {
    const testInputSpy = jest.fn();
    const testInput = inputValue => {
      testInputSpy();
      return Regex.email.test(inputValue)
    };
    await testTextInput({ testInput, value: "johndoe@emai", status: 5 }, testInputSpy);
    await testTextInput({ testInput, value: "johndoe@email.com", status: 6 }, testInputSpy);
  });

  it("calls the onChange action when regex test is valid", async () => {
    await testTextInput({ testInput: Regex.email ,value: "johndoe@emai", status: 5});
    await testTextInput({ testInput: Regex.email ,value: "johndoe@email.com", status: 6 });
  });

  it("calls the onChange action when regex name test is valid", async () => {
    await testTextInput({ testInput: "email", value: "johndoe@emai", status: 5 });
    await testTextInput({ testInput: "email", value: "johndoe@email.com", status: 6 });
  });

  it("focus on input successfully", async () => {
    const { wrapper } = setup(null, true);
    const spyOnInputFocus = jest.spyOn(wrapper.instance().textInputRef.current, 'focus');
    wrapper.instance().focus();
    expect(spyOnInputFocus).toHaveBeenCalled();
  });

  it("sets the state of the input as valid", async () => testSetStatusMethod({ params: ['success:fixed'], statusNo: 6 }));

  it("sets the state of the input as invalid", async () => testSetStatusMethod({ params: ['fail:fixed', 'Invalid input'], statusNo: 5 }));

  it("sets the state of the input with invalid status", async () => testSetStatusMethod({ params: ['correct:fixed' ], statusNo: 0 }));

});
