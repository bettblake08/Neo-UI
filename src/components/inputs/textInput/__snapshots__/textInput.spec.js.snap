// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput  renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextInput
    componentStateKey="neoComponents"
    config={
      Object {
        "floatingLabel": true,
        "label": "Name",
        "length": 30,
        "placeholder": "Brian Bett",
        "text": "",
        "type": "neo-text-input",
      }
    }
    defaultStatus={0}
    name="Name"
    parent={
      Object {
        "setState": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
        "state": Object {},
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="neo-text-input__label neo-font--input"
        >
          Name
        </div>,
        <input
          className="neo-font--input"
          id="Name"
          onChange={[Function]}
          placeholder="Brian Bett"
          type="text"
          value=""
        />,
        <label
          htmlFor="Name"
        >
          Name
        </label>,
        <div
          className="neo-font--comment neo-text-input__comment--active"
        />,
        <div
          className="neo-font--comment neo-text-input__error--disabled"
        >
          
        </div>,
      ],
      "className": "neo-text-input--normal neo-font--input has-float-label",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Name",
          "className": "neo-text-input__label neo-font--input",
        },
        "ref": null,
        "rendered": "Name",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "neo-font--input",
          "id": "Name",
          "onChange": [Function],
          "placeholder": "Brian Bett",
          "type": "text",
          "value": "",
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Name",
          "htmlFor": "Name",
        },
        "ref": null,
        "rendered": "Name",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "neo-font--comment neo-text-input__comment--active",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "neo-font--comment neo-text-input__error--disabled",
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="neo-text-input__label neo-font--input"
          >
            Name
          </div>,
          <input
            className="neo-font--input"
            id="Name"
            onChange={[Function]}
            placeholder="Brian Bett"
            type="text"
            value=""
          />,
          <label
            htmlFor="Name"
          >
            Name
          </label>,
          <div
            className="neo-font--comment neo-text-input__comment--active"
          />,
          <div
            className="neo-font--comment neo-text-input__error--disabled"
          >
            
          </div>,
        ],
        "className": "neo-text-input--normal neo-font--input has-float-label",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Name",
            "className": "neo-text-input__label neo-font--input",
          },
          "ref": null,
          "rendered": "Name",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "neo-font--input",
            "id": "Name",
            "onChange": [Function],
            "placeholder": "Brian Bett",
            "type": "text",
            "value": "",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Name",
            "htmlFor": "Name",
          },
          "ref": null,
          "rendered": "Name",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "neo-font--comment neo-text-input__comment--active",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "neo-font--comment neo-text-input__error--disabled",
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
